Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerVC.swift

```diff
@@ -343,63 +359,37 @@ open class _ChatMessageComposerVC<ExtraData: ExtraDataTypes>: _ViewController,
 
     // MARK: Attachments
 
-    public enum SelectedAttachments {
-        case media, documents
-    }
-    
-    open var selectedAttachments: SelectedAttachments? {
-        if imageAttachments.isEmpty, documentAttachments.isEmpty {
-            return .none
-        } else {
-            return imageAttachments.isEmpty ? .documents : .media
-        }
-    }
-    
-    open var imageAttachments: [URL] = [] {
-        didSet {
-            didUpdateImageAttachments()
-        }
-    }
-    
-    open var documentAttachments: [URL] = [] {
+    open var urlAttachments: [URL] = [] {
         didSet {
-            didUpdateDocumentAttachments()
+            didUpdateAttachments()
         }
     }
     
     open var attachments: [AttachmentEnvelope] {
-        switch selectedAttachments {
-        case .media:
-            return imageAttachments.compactMap(AttachmentEnvelope.init)
-        case .documents:
-            return documentAttachments.compactMap(AttachmentEnvelope.init)
-        case .none:
-            return []
-        }
-    }
-
-    func didUpdateImageAttachments() {
-        composerView.inputMessageView.imageAttachmentsView.content = imageAttachments.compactMap {
-            guard let preview = UIImage(contentsOfFile: $0.path) else { return nil }
-            return ImageAttachmentPreview(image: preview)
-        }
-        Animate {
-            self.composerView.inputMessageView.imageAttachmentsView.isHidden = self.imageAttachments.isEmpty
-        }
-        updateSendButton()
+        urlAttachments.compactMap(AttachmentEnvelope.init)
     }
     
-    func didUpdateDocumentAttachments() {
-        composerView.inputMessageView.documentAttachmentsView.documents = documentAttachments.map {
-            let filePreview = appearance.images.documentPreviews[$0.pathExtension]
-            return (
-                filePreview ?? appearance.images.fileFallback,
-                $0.lastPathComponent,
-                $0.attachmentFile?.size ?? 0
-            )
+    open var selectedAttachmentType: AttachmentType?
+
+    open func didUpdateAttachments() {
+        attachmentsViewController.content = attachments.compactMap {
+            switch $0.payload {
+            case let payload as AttachmentImagePayload:
+                guard let preview = UIImage(contentsOfFile: payload.imagePreviewURL.path) else { return nil }
+                return ImageAttachmentPreview(image: preview)
+                
+            case let payload as AttachmentFilePayload:
+                let filePreview = $0.localFileURL.flatMap { appearance.images.documentPreviews[$0.pathExtension] }
+                    ?? appearance.images.fileFallback
+                return FileAttachmentPreview(image: filePreview, name: payload.title, size: payload.file.size)
+
+            default:
+                log.error("Unexpected attachment payload type.")
+                return nil
+            }
         }
```

Comment:
What if instead of this created something like:
```
protocol AttachmentPreview { 
  func previewView() -> UIView // or UIView? or even AttachmentPreviewView?
}
```
and said that `open var attachments: [AttachmentEnvelope & AttachmentPreview]`? Customers will be able to implement this protocol and provide their own implementation for this (i.e. map preview, etc.).

Reply:
Done, please check